File: 백준/Gold/17298. 오큰수/오큰수.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.Stack;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        Stack<Integer> s = new Stack<>();
        int N = Integer.parseInt(br.readLine());
        int[] answer = new int[N];
        int[] nums = new int[N];
        Arrays.fill(answer, -1);
        StringTokenizer st = new StringTokenizer(br.readLine());
        for (int i = 0; i < N; i++) {
            nums[i] = Integer.parseInt(st.nextToken());
        }

        for (int i = N - 1; i >= 0; i--) {
            while (!s.isEmpty() && s.peek() <= nums[i]) {
                s.pop();
            }

            if (!s.isEmpty()) {
                answer[i] = s.peek();
            }
            s.push(nums[i]);
        }

        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < answer.length; i++) {
            sb.append(answer[i]).append(" ");
        }

        System.out.println(sb);
    }
}

File: 백준/Gold/17298. 오큰수/오큰수.py


N = int(input())
A = [int(s) for s in input().split()]
answer = [-1] * N
stack = []
for i in range(N - 1, -1, -1):
    while stack and A[i] >= A[stack[-1]]:
        stack.pop()
    if stack and A[stack[-1]] > A[i]:
        answer[i] = A[stack[-1]]
    stack.append(i)

print(*answer)